Create poll function for handel aggregations

nibhar opened the pull request:
This PR implements a poll function for handel.

Other smaller additions have been made like removing the tokio dependence and some changes to lock scoping as well as wakers. 

Lastly the test has been updated to be more reliable.

While this PR may affect #2623 I do not expect it to have a big impact.

nibhar committed:
Handel: Remove unnecessary waker

nibhar committed:
Handel: Remove tokio dependency

nibhar committed:
Handel: Update test to not fanout messages

nibhar committed:
Adapt the validator to use the handel verifier instead of protocol

nibhar committed:
Handel: Wait for aggregations to finish instead of sleeping

nibhar committed:
Handel: Improve lock scoping

nibhar committed:
move handel dependency to dev dependencies in the primitives crate

nibhar committed:
Add Identifier trait bounds for Handel

nibhar committed:
Remove unused individual_received from Handel's store

nibhar committed:
Remove TODO comment as it is not necessary

nibhar committed:
Add additional early return to check_merge, add identity to return

ii-cruz commented on handel/src/aggregation.rs:
```diff
@@ -321,105 +314,194 @@ impl<
+        );
+
+        // in case the level of this todo has not started, start it now as we have already contributions on it.
+        self.start_level(todo.level, &*store);
```
We do not check if the level has been started so we might be repeatedly starting the level.

ii-cruz replied:
my bad this is a none issue

ii-cruz commented on handel/src/aggregation.rs:
```diff
@@ -321,105 +314,194 @@ impl<
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Check if the automatic update interval triggers, if so perform the update.
+        if let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```
By increasing this timeout, we might be falling into these periodic updates more often and thus increase the number of messages sent.

ii-cruz replied:
Talked with @nibhar and both timeouts were increased so this is not relevant

ii-cruz commented on handel/src/aggregation.rs:
```diff
@@ -321,105 +314,194 @@ impl<
+        }
+
+        // Poll the verification future if there is one.
+        let mut best_aggregate = if let Some(verification_future) = &mut self.current_verification {
```
why is the verification after the intervals and timeouts? Shouldn't it be the other way around?

nibhar replied:
I changed the ordering such that the timers are polled last, meaning that any updates created in the potential subsequent calls will already be able to use any new aggregate produced by a previous signature being added.

nibhar committed:
Handel: Change order within poll function

This is done such that a new best aggregate is produced first. If updates are produced later they can make use of the already existing better update.

hrxi commented on handel/src/todo.rs:
```diff
@@ -102,22 +99,15 @@ where
         }
     }
 
+    /// Safe without a wake as it is only called in the constructor of Aggregation.
```
Could you add a member variable that tracks whether it has ever been polled, and add an assert here that this hasn't happened? This would ensure that this is robust in case of future code modification.

nibhar replied:
Good call! I add it.

hrxi commented on handel/src/protocol.rs:
```diff
@@ -9,11 +9,15 @@ use crate::{
+    verifier::Verifier,
 };
 
 #[async_trait]
```
The `#[async_trait]` can go, this trait no longer contains any `async` functions.

hrxi commented on handel/src/evaluator.rs:
```diff
@@ -256,13 +247,18 @@ where
+            return weight == self.partitioner.size();
         }
-        let range = self.partitioner.range(level);
+        let range = self.partitioner.range(msg.level as usize);
```
The level should probably not be passed as an `usize` (it has nothing to do with the host architecture's pointer bit width), but I guess it doesn't have to be in this pull request.

nibhar replied:
I believe that is going to be the case in a couple more places. Would you say this would be more of a cosmetic change or do you see a more functional benefit?

I would do it later if it is considered beneficial.

hrxi replied:
Cosmetic. Our levels don't get large enough to get in the range of the platform differences.

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -326,100 +332,190 @@ impl<
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Check if the automatic update interval triggers, if so perform the update.
+        if let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```
```suggestion
        while let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```
This needs to be a `while let`, otherwise the interval doesn't register itself with the waker if `Poll::Ready` is returned: https://docs.rs/tokio/1.38.0/tokio/time/struct.Interval.html#method.poll_tick.

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -326,100 +332,190 @@ impl<
+            self.automatic_update();
+        }
+
+        if let Poll::Ready(_instant) = self.start_level_interval.as_mut().poll_tick(cx) {
```
```suggestion
        while let Poll::Ready(_instant) = self.start_level_interval.as_mut().poll_tick(cx) {
```

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -326,100 +332,190 @@ impl<
 
-    fn into_inner(self) -> (LevelUpdateStream<P, TId>, LevelUpdateSender<N>) {
-        (self.todos.into_stream(), self.sender)
+        // Poll sender to until it is pending sending as many messages as possible.
```
```suggestion
        // Poll sender until it is pending, sending as many messages as possible.
```

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -326,100 +332,190 @@ impl<
-    fn into_inner(self) -> (LevelUpdateStream<P, TId>, LevelUpdateSender<N>) {
-        (self.todos.into_stream(), self.sender)
+        // Poll sender to until it is pending sending as many messages as possible.
+        while let Poll::Ready(t) = self.sender.poll_next_unpin(cx) {
+            t.expect("Sender should never return None");
+        }
```
It seems that the `LevelUpdateSender` can never return `Poll::Ready`

```suggestion
        assert!(self.sender.poll_next_unpin(cx).is_pending());
```

Since that is the case, maybe it shouldn't implement `Stream`, but rather `Future`.

nibhar replied:
Good find! I made it a Future instead. I also added the asserts.

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -430,98 +526,132 @@ impl<
 
-        Poll::Ready((self.level_update.aggregate.clone(), None))
+        // Send as many messages as possible.
+        let _ = self.sender.poll_next_unpin(cx);
```
```suggestion
        assert!(self.sender.poll_next_unpin(cx).is_pending());
```

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -430,98 +526,132 @@ impl<
+    /// The aggregation is currently ongoing and it is going to produce stream items as well as network traffic.
+    Ongoing(OngoingAggregation<TId, P, N>),
+    /// The aggregation is finished and it will no longer produce stream items and it will reduce network traffic
+    /// to answering messages with an unfinished aggregate within them
```
```suggestion
    /// to answering messages with an unfinished aggregate within them.
```

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -430,98 +526,132 @@ impl<
+            Self::Transitioning => {
+                unreachable!("Aggregation should never be transitioning when polled.")
+            }
+            Self::Finished(ref mut finished_aggregation) => {
```
```suggestion
            Self::Finished(finished_aggregation) => {
```
Here and below, I thought the `ref mut` is unnecessary, but maybe I'm wrong.

nibhar replied:
Works 👍!

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -430,98 +526,132 @@ impl<
+            }
+            Self::Finished(ref mut finished_aggregation) => {
+                // Finished aggregations are simply polled as the future will never produce a value.
+                return finished_aggregation.poll_unpin(cx);
```
```suggestion
                assert!(finished_aggregation.poll_unpin(cx).is_pending());
                return Poll::Pending;
```

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -430,98 +526,132 @@ impl<
+                // Finished aggregations are simply polled as the future will never produce a value.
+                return finished_aggregation.poll_unpin(cx);
+            }
+            Self::Ongoing(ref mut ongoing_aggregation) => {
```
```suggestion
            Self::Ongoing(ongoing_aggregation) => {
```
(maybe incorrect.)

hrxi commented on handel/Cargo.toml:
```diff
@@ -22,8 +22,6 @@ parking_lot = "0.12"
 serde = "1.0"
 thiserror = "1.0"
-tokio = { version = "1.38", features = ["macros", "rt", "time"] }
-tokio-stream = "0.1"
```
🎉

hrxi commented on network-mock/src/lib.rs:
```diff
@@ -31,7 +31,7 @@ pub struct MockAddress(u64);
     Into,
 )]
-pub struct MockPeerId(u64);
+pub struct MockPeerId(pub u64);
```
The `pub` seems unnecessary in this commit.
```suggestion
pub struct MockPeerId(u64);
```

nibhar replied:
The test uses it to create a specific Peer to send a message to. I think without it being pub it cannot be used like so:
https://github.com/nimiq/core-rs-albatross/pull/2693/files#diff-026455fd64af083e01fb8f83c0d5430cd7e37ba6096fa32800f3a08a022f2deeR192

hrxi replied:
Hmmm. That test kinda confuses `MockPeerId`s and handel's IDs. There's probably a mapping between them somewhere, no?

hrxi replied:
Ah, it manually sets up the peer IDs for that to work.

You can go from `u64` to `MockPeerId` using `MockPeerId::from`.

hrxi commented on handel/tests/mod.rs:
```diff
@@ -330,7 +338,14 @@ async fn it_can_aggregate() {
     net.disconnect();
 
     // give the other aggregations time to complete themselves
-    nimiq_time::sleep(Duration::from_millis(100)).await;
```
🎉

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -327,8 +314,11 @@ where
+
             // finally start the level.
-            self.start_level(level);
+            self.start_level(level, &*store);
```
```suggestion
            self.start_level(level, &store);
```
The dereference is likely unnecessary.

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -346,30 +336,26 @@ where
 
         // in case the level of this todo has not started, start it now as we have already contributions on it.
-        self.start_level(todo.level);
+        self.start_level(todo.level, &*store);
```
```suggestion
        self.start_level(todo.level, &store);
```

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -346,30 +336,26 @@ where
+        self.start_level(todo.level, &*store);
         // check if a level was completed by the addition of the contribution
-        self.check_completed_level(todo.level);
+        self.check_completed_level(todo.level, &*store);
```
```suggestion
        self.check_completed_level(todo.level, &store);
```

hrxi commented on primitives/block/Cargo.toml:
```diff
@@ -43,5 +42,6 @@ nimiq-utils = { workspace = true, features = ["merkle"] }
 nimiq-vrf = { workspace = true, features = ["serde-derive"] }
 
 [dev-dependencies]
+nimiq-handel = { workspace = true }
```
🎉

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -435,6 +423,18 @@ where
         };
 
+        // Check if the automatic update interval triggers, if so perform the update.
+        if let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```
```suggestion
        while let Poll::Ready(_instant) = self.periodic_update_interval.as_mut().poll_tick(cx) {
```

nibhar replied:
I think these are duplicates with an above comment from a prior commit?

hrxi commented on handel/src/aggregation.rs:
```diff
@@ -435,6 +423,18 @@ where
+            self.automatic_update();
+        }
+
+        if let Poll::Ready(_instant) = self.start_level_interval.as_mut().poll_tick(cx) {
```
```suggestion
        while let Poll::Ready(_instant) = self.start_level_interval.as_mut().poll_tick(cx) {
```

nibhar replied:
I think these are duplicates with an above comment from a prior commit?

hrxi reviewed (commented):
The general changes look very good to me. :)

I found a couple of minor issues.

nibhar committed:
Address review comments

nibhar committed:
Handel: Implement a proper poll function

hrxi reviewed (approved):
Nice changes! They look good to me now. :)
